### PR TITLE
Add assert for ParticleHandler when particles to be inserted are not found

### PR DESCRIPTION
**Pull request by masterleinad**
Similar to #12221.

### CODE REVIEW

File: source/particles/particle_handler.cc [hunk 1]
```diff
@@ -437,11 +439,19 @@ namespace Particles
...
+      GridTools::compute_point_locations_try_all(*triangulation_cache,
+                                                 positions);
 
     auto &cells           = std::get<0>(point_locations);
     auto &local_positions = std::get<1>(point_locations);
     auto &index_map       = std::get<2>(point_locations);
+    auto &missing_points  = std::get<3>(point_locations);
+    // If a point was not found, throwing an error, as the old
+    // implementation of compute_point_locations would have done
+    AssertThrow(std::get<3>(point_locations).size() == 0,
```
Comment (marcfehling) [inline source/particles/particle_handler.cc:451]:
```suggestion
    AssertThrow(missing_points.size() == 0,
```
I guess this was the purpose of introducing the reference above?

**Review by marcfehling — APPROVED** [review]

File: source/particles/particle_handler.cc (hunk 1 above)
Comment (masterleinad) [reply source/particles/particle_handler.cc]: Sure.

**Comment by Rombur** [comment]
```
The following tests FAILED:

	241 - particles/particle_handler_serial_07.debug (Failed)
```

**Comment by masterleinad** [comment]
> The following tests FAILED:
>	241 - particles/particle_handler_serial_07.debug (Failed)

That's essentially https://github.com/dealii/dealii/pull/11276/files?w=1#r629301779. @luca-heltai Any suggestions on what to do here? Should we say that this function is not expected to work in the co-dimension 1 case?

**Comment by luca-heltai** [comment]
Yes. I think this is correct: this function is not expected to work in the co-dimension 1 case. The transform real to unit method in the co-dimension 1 case actually performs a projection to the unit cell. We need to add some checks (and tolerances) for these to be reliable.

**Comment by kronbichler** [comment]
The test `particle_handler_serial_07.debug` is still failing.

**Comment by masterleinad** [comment]
> The test `particle_handler_serial_07.debug` is still failing.

Should work now.

**Review by bangerth — COMMENTED** [review]
Does that mean that we can no longer use `ParticleHandler` for the codim-1 case? Was it ever usable before?

**Comment by masterleinad** [comment]
> Does that mean that we can no longer use `ParticleHandler` for the codim-1 case? Was it ever usable before?

Per https://github.com/dealii/dealii/pull/11276#discussion_r629301779 by @luca-heltai:
> Yes. The previous algorithm was returning garbage. Particles don't work (yet) properly on manifolds of co-dimension one or two, since in those cases the transformation from real to unit cell is not invertible in general, but just on the manifold, and our algorithms are not robust for this case. The current algorithm won't find particles because it starts off by checking inside a bounding box of the cell first, and all particles here are away from the manifold.

File: source/particles/particle_handler.cc [hunk 2]
```diff
@@ -437,11 +439,19 @@ namespace Particles
...
+                                                 positions);
 
     auto &cells           = std::get<0>(point_locations);
     auto &local_positions = std::get<1>(point_locations);
     auto &index_map       = std::get<2>(point_locations);
+    auto &missing_points  = std::get<3>(point_locations);
+    // If a point was not found, throwing an error, as the old
+    // implementation of compute_point_locations would have done
+    AssertThrow(missing_points.size() == 0,
+                VectorTools::ExcPointNotAvailableHere());
```
Comment (kronbichler) [inline source/particles/particle_handler.cc:452]: We are throwing a different exception now, but given that we remove the other one on purpose in #11276, this is fine.

**Review by kronbichler — APPROVED** [review]

**Comment by luca-heltai** [comment]
If I recall correctly, @gassmoeller had this on porpuse to ignore particles that don't fall within the domain, and instead call a signal with the particles that are out, rather than just throwing an exception. Wouldn't it be better?

**Pull request merged**